(Issue #87255) TRUNCATE on S3 table engine deletes all objects under table prefix

Seth-Banker opened the pull request:
### Changelog category
- Bug Fix

### Changelog entry
Fix TRUNCATE for S3 table engine so that it deletes all objects under the table prefix.
Previously TRUNCATE only removed the configured paths. Data files created by inserts could remain after a server restart or with hive partitions. Now TRUNCATE lists the read prefix and removes all objects, matching TRUNCATE semantics in other engines.

### Documentation entry for user-facing changes
TRUNCATE TABLE now fully empties S3 backed tables by deleting all objects under the configured prefix.

**Motivation:**  
Users expect TRUNCATE to remove all data while preserving the table schema. With S3 tables this was previously incomplete.

**Behavior:**  
- Works for both non-partitioned and hive-partitioned layouts.  
- Globs and archive paths remain read-only and cannot be truncated.  

**Example:**
```sql
CREATE TABLE s3_tbl
(
  name String,
  value UInt32
)
ENGINE = S3('s3://my-bucket/path/{partition}/data_*.parquet', 'Parquet');

INSERT INTO s3_tbl VALUES ('x', 1), ('y', 2);
TRUNCATE TABLE s3_tbl;
SELECT count() FROM s3_tbl; -- returns 0
```

Seth-Banker committed:
Update StorageObjectStorage.cpp

Fix truncate for S3 table engine to remove all objects under table prefix.

Seth-Banker committed:
Update StorageObjectStorage.h

Added missing headers needed for changes.

Seth-Banker committed:
Update StorageObjectStorage.h

Switching changes from header to cpp file.

Seth-Banker committed:
Update StorageObjectStorage.cpp

Added missing header includes.

Seth-Banker commented:
Included proper header files. This can be tested again with new changes.

Seth-Banker committed:
Update StorageObjectStorage.cpp

Unfortunate typo.

Seth-Banker committed:
Update StorageObjectStorage.cpp

Seth-Banker committed:
Update StorageObjectStorage.cpp

Wrong references were used. Objects should be handled properly in truncate()

Seth-Banker commented:
The [failed job](https://s3.amazonaws.com/clickhouse-test-reports/PRs/87260/be6c1c0d6dee0d58d7488941dd6a9fb9ddec7970//install_packages_amd_debug/install_clickhouse_binary_in_deb.log) doesn't have enough memory allocated to install.

Seth-Banker committed:
Force CI Re-Run

no real changes. error in automated test docker.

Seth-Banker committed:
Force CI re-run

Quality check process does not work independently and I don't want to wait another 24 hours for a code reviewer.

Seth-Banker commented:
@rienath

Please re-run all skipped/failed processes.  

The environments that are trying to test this keep running into errors. I have not had 1 full run without an environment having some internal error. 

Would it be possible to just have someone manually review this? I don't see much point in fiddling with a faulty CI.

Seth-Banker committed:
Third Force CI re-run.

CI has failed due to environment errors 3 times.

Seth-Banker commented:
@rienath 

Please re-run failed tasks. I'm assuming skipped tasks need to be re-ran to.

I tried re-running by submitting inconsequential gits but the entire process constantly requires restarting failed tasks. I have not had 1 full run without a testing environment having some sort of internal error.

Would it just be easier to have someone manually review this?

Seth-Banker commented:
@rienath Hey just wanted to check in with you. Process failing due to environment error. 

Was wondering if you wanted to re-run the skipped/failed tests? Or would a manual review be easier?

rienath commented:
@Seth-Banker let's leave it up to the reviewer to decide. FYI: to pass CI with bugfix changelog entry, you need to introduce a test fails on master, but works on your branch

arthurpassos reviewed (commented):
Can you also add a few stateless tests?

1. No partitioning
2. Wildcard partition strategy
3. Hive partition strategy

As simple as `insert -> select -> truncate -> select`.

If you want to cover it a bit more, you could also add an integration test in which you re-start the server to force the in-memory `paths` list to disappear and see how your implementation behaves

Seth-Banker commented:
> Can you also add a few stateless tests?
> 
>     1. No partitioning
> 
>     2. Wildcard partition strategy
> 
>     3. Hive partition strategy
> 
> 
> As simple as `insert -> select -> truncate -> select`.
> 
> If you want to cover it a bit more, you could also add an integration test in which you re-start the server to force the in-memory `paths` list to disappear and see how your implementation behaves

Of course, planned on doing it today but finishing some school work. Will probably produce tomorrow PST.  

Thank you so much.

Sorry if the commits are a bit odd. I'm new to contributing to other projects, I'm sure you could already tell that by now though.

Seth-Banker committed:
Update StorageObjectStorage.cpp

Implemented suggestions made by @arthurpassos

Seth-Banker committed:
Same as previous

Comment edit on truncate() in StorageObjectStorage.cpp didn't go through for some reason.

Seth-Banker committed:
Update StorageObjectStorage.cpp

Removed trailing whitespace.

Seth-Banker committed:
Update StorageObjectStorage.cpp

raw_path references previously weren't updated in truncate()

Seth-Banker commented:
> Can you also add a few stateless tests?
> 
>     1. No partitioning
> 
>     2. Wildcard partition strategy
> 
>     3. Hive partition strategy
> 
> 
> As simple as `insert -> select -> truncate -> select`.
> 
> If you want to cover it a bit more, you could also add an integration test in which you re-start the server to force the in-memory `paths` list to disappear and see how your implementation behaves

@arthurpassos Test shown in original PR post and recommended changes added to repo.
